File: AIDL/src/main/java/com/example/liubin/myapplication/MainActivity.java
package com.example.liubin.myapplication;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.Toast;

import java.util.List;

public class MainActivity extends AppCompatActivity {
    private Button bindBtn;
    private Button greetBtn;
    private Button unbindBtn;

    private IGreetService greetService;
    private ServiceConnection conn = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.i("ServiceConnection", "onServiceConnected() called");
            greetService = IGreetService.Stub.asInterface(service);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            //This is called when the connection with the service has been unexpectedly disconnected,
            //that is, its process crashed. Because it is running in our same process, we should never see this happen.
            Log.i("ServiceConnection", "onServiceDisconnected() called");
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });

        // 绑定按钮
        bindBtn = (Button) findViewById(R.id.button1);
        bindBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent("android.intent.action.AIDLService");
                intent.setPackage(getPackageName());
                bindService(getExplicitIntent(MainActivity.this, intent), conn, Context.BIND_AUTO_CREATE);

                bindBtn.setEnabled(false);
                greetBtn.setEnabled(true);
                unbindBtn.setEnabled(true);
            }
        });

        // 调用按钮
        greetBtn = (Button) findViewById(R.id.button2);
        greetBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    Person person = new Person();
                    person.setName("scott");
                    person.setSex(0);
                    String retVal = greetService.greet(person);
                    Toast.makeText(MainActivity.this, retVal, Toast.LENGTH_SHORT).show();

                    person = greetService.getPerson(person);
                    Log.e(person.getName(), "" + person.getSex());
                } catch (RemoteException e) {
                    Toast.makeText(MainActivity.this, "error", Toast.LENGTH_SHORT).show();
                }
            }
        });

        // 取消绑定按钮
        unbindBtn = (Button) findViewById(R.id.button3);
        unbindBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                unbindService(conn);

                bindBtn.setEnabled(true);
                greetBtn.setEnabled(false);
                unbindBtn.setEnabled(false);
            }
        });

        bindBtn.setEnabled(true);
        greetBtn.setEnabled(false);
        unbindBtn.setEnabled(false);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public static Intent getExplicitIntent(Context context, Intent implicitIntent) {
        // Retrieve all services that can match the given intent
        PackageManager pm = context.getPackageManager();
        List<ResolveInfo> resolveInfo = pm.queryIntentServices(implicitIntent, 0);
        // Make sure only one match was found
        if (resolveInfo == null || resolveInfo.size() != 1) {
            return null;
        }
        // Get component info and create ComponentName
        ResolveInfo serviceInfo = resolveInfo.get(0);
        String packageName = serviceInfo.serviceInfo.packageName;
        String className = serviceInfo.serviceInfo.name;
        ComponentName component = new ComponentName(packageName, className);
        // Create a new intent. Use the old one for extras and such reuse
        Intent explicitIntent = new Intent(implicitIntent);
        // Set the component to be explicit
        explicitIntent.setComponent(component);
        return explicitIntent;
    }
}

File: Kotlin/src/main/java/com/example/kotlin/MainActivity.kt
package com.example.kotlin

import android.app.Activity
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

  private var age: Int? = null
  private var mText: TextView? = null;
  private val mData: List<String> = ArrayList();

  // 扩展方法
  fun Activity.toast(message: String? = "") {
    Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
  }

  // 类似单例模式
  object Status {
    private val success: Int = 1
    public fun say(): Int {
      System.out.println(success)
      return success
    }
  }

  //常量定义
  private val PAGE_SIZE = 20

  companion object {
    private val ITEM_TYPE_DATA = 1
    private val ITEM_TYPE_HEADER = 2
    private val ITEM_TYPE_FOOTER = 3
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    mData.forEach { s ->
      toast(s)
    }

    for (s: String in mData) {
      toast(s)
    }

    // 常量使用
    Companion.ITEM_TYPE_DATA

    // 点击事件
    text?.setOnClickListener {
      var age: Int = 23
      var name: String = "name"
    }

    text.setOnClickListener({
      toast("hahah")
      vars(1, 2, 3, 4, 5)  // 输出12345
    })

    text?.setOnClickListener(View.OnClickListener { v ->
      (v as TextView).text = "hah"

    })

    text.setOnClickListener({ v ->
      (v as TextView).text = "点击了哦"
      toast("hahah")
    });

    text.setOnClickListener(object : View.OnClickListener {

      internal var age = 23

      private fun getAge() {
        return
      }

      override fun onClick(v: View?) {
        (v as TextView).text = "点击了哦"
        toast("hahah")
      }

    })

    val p: Person = Person(12, "liubin")
    p.setAddress("dizhi")
    toast(p.getMessage())
    toast(p.getAddress())
  }

  // 可变参数
  fun vars(vararg v: Int) {
    for (vt in v) {
      print(vt)
    }
  }


}
